deepmd/dpmodel/descriptor/dpa4_nn/norm.py

```
@@ -0,0 +1,647 @@
+        self.trainable = bool(trainable)
+        prec = PRECISION_DICT[self.precision.lower()]
+
+        self.degree_index_m = np.asarray(degree_index_m, dtype=np.int64)
+
+        # Pre-fuse degree balancing and channel averaging into a single weight:
+        #   w_d = 1 / (n_coeff_l * (lmax+1) * C)
+        # where n_coeff_l is the number of retained coefficients for degree l in
+        # the reduced layout.
+        weights = np.zeros(self.degree_index_m.size, dtype=prec)
+        scale = 1.0 / ((self.lmax + 1) * self.channels)
+        for l in range(self.lmax + 1):
```

Ruff (0.15.15)

[error] Ambiguous variable name: `l`

(E741)